doc/cookbook/namedRoutes/NamedRoutes.lhs

```diff
@@ -0,0 +1,296 @@
+    go (m:ms) = go ms
+
+updateMovieHandler :: MovieId -> Movie -> Handler NoContent
+updateMovieHandler requestedMovieId newMovie =
+   -- update the movie list in the database...
+   pure NoContent
+
+deleteMovieHandler :: MovieId -> Handler NoContent
+deleteMovieHandler _ =
+   -- delete the movie from the database...
+   pure NoContent
+
+```
+
+And assemble them together with the record structure, which is the glue here.
+
+```haskell
+server ::  API AsServer
+server =
+  API
+    { version = versionHandler
+    , movies = moviesHandler
+    }
+
+moviesHandler :: MoviesAPI AsServer
+moviesHandler =
+  MoviesAPI
+    { list = movieListHandler
+    , movie = movieHandler
+    }
+
+movieHandler :: MovieId -> MovieAPI AsServer
+movieHandler movieId = MovieAPI
+    { get = getMovieHandler movieId
+    , update = updateMovieHandler movieId
+    , delete = deleteMovieHandler movieId
+    }
+```
+As you might have noticed, we build our handlers out of the same record types we used to define our API: `MoviesAPI` and `MovieAPI`. What kind of magic is this ?
+
+Finally, we can run the server and connect the API routes to the handlers as usual:
+
+``` haskell
+api :: Proxy MovieCatalogAPI
+api = Proxy
+
+main :: IO ()
+main = run 8081 app
+
+app :: Application
+app = serve api server
+
+```
+Yay! That’s done and we’ve got our server!
+
+## The Client
+
+The client, so to speak, is very easy to implement:
+
+``` haskell
+movieCatalogClient :: API (AsClientT ClientM)
+movieCatalogClient = client api -- remember: api :: Proxy MovieCatalogAPI
+```
+
+We’ve also introduced some operators that help navigate through the nested records.
+
+`(//)` is used to jump from one record to another.
+`(/:)` is used to provide a parameter, whether it be a query param or a capture.
+
+Let’s use those nice helpers for our movie catalog:
+
+```haskell
+listMovies :: Maybe SortBy -> ClientM [Movie]
+listMovies sortBy = movieCatalogClient // movies // list /: sortBy
+
+getMovie :: MovieId -> ClientM (Maybe Movie)
+getMovie movieId = movieCatalogClient // movies // movie /: movieId // get
+
+updateMovie :: MovieId -> Movie -> ClientM NoContent
+updateMovie movieId newMovie = movieCatalogClient // movies // movie /: movieId // update /: newMovie
+
+deleteMovie :: MovieId -> ClientM NoContent
+deleteMovie movieId = movieCatalogClient // movies // movie /: movieId // delete
+```
+
+Done! We’ve got our client!
+
+## Conclusion
+
+We hope that you found this cookbook helpful, and that you now feel more confident using the record-based APIs, nested or not.
+
+If you are interested in further understanding the built-in Servant combinators, see [Structuring APIs](../structuring-apis/StructuringApis.html).
+
+Since `NamedRoutes` is based on the Generic mechanism, you might want to have a look at [Sandy Maguire’s _Thinking with Types_ book](https://doku.pub/download/sandy-maguire-thinking-with-typesz-liborgpdf-4lo5ne7kdj0x).
```

Comment:
Not nice to put a piracy link. Better link the official website: https://thinkingwithtypes.com/